File: IMU_Reader.py
import serial
import re


def parse_serial(serial_msg):
    # Regex pattern to match the formatted line from the Arduino
    pattern = (r'Mag x:([-+]?[0-9]*\.?[0-9]+) y:([-+]?[0-9]*\.?[0-9]+) z:([-+]?[0-9]*\.?[0-9]+) \| Gyro x:([-+]?['
               r'0-9]*\.?[0-9]+) y:([-+]?[0-9]*\.?[0-9]+) z:([-+]?[0-9]*\.?[0-9]+) \| Acc x:([-+]?[0-9]*\.?[0-9]+) '
               r'y:([-+]?[0-9]*\.?[0-9]+) z:([-+]?[0-9]*\.?[0-9]+)')
    match = re.search(pattern, serial_msg)
    if match:
        # Extracting all matched groups into a flat list of floats
        data = [float(match.group(i)) for i in range(1, 10)]
        return {
            "Mag": data[0:3],  # First 3 values for magnetometer
            "Gyro": data[3:6],  # Next 3 values for gyroscope
            "Acc": data[6:9]  # Last 3 values for accelerometer
        }
    return None


# Assuming '/dev/ttyACM0' is the correct port and 9600 is the baud rate
ser = serial.Serial('/dev/ttyACM0', 9600, timeout=1)

# Reading a single line from the serial port
serial_msg_bytes = ser.readline()
serial_msg = serial_msg_bytes.decode(encoding='utf-8').strip()

# Parsing the line to extract IMU data
imu_data = parse_serial(serial_msg)
if imu_data:
    print(f"Magnetometer: {imu_data['Mag']}")
    print(f"Gyroscope: {imu_data['Gyro']}")
    print(f"Accelerometer: {imu_data['Acc']}")
else:
    print("Failed to parse IMU data.")

File: IMU_Reader.cpp
#include "Arduino_BMI270_BMM150.h"

void setup() {
  Serial.begin(9600);
  while (!Serial); // Wait for the serial port to connect - necessary for Leonardo/Micro
  Serial.println("Started");

  if (!IMU.begin()) {
    Serial.println("Failed to initialize IMU!");
    while (1); // Infinite loop to halt operation
  }
}

void loop() {
  float mx, my, mz; // Magnetometer values
  float gx, gy, gz; // Gyroscope values
  float ax, ay, az; // Accelerometer values

  bool mAvailable = IMU.magneticFieldAvailable();
  bool gAvailable = IMU.gyroscopeAvailable();
  bool aAvailable = IMU.accelerationAvailable();

  if (mAvailable) {
    IMU.readMagneticField(mx, my, mz);
  }

  if (gAvailable) {
    IMU.readGyroscope(gx, gy, gz);
  }

  if (aAvailable) {
    IMU.readAcceleration(ax, ay, az);
  }

  // Now print all available data in one line
  if (mAvailable || gAvailable || aAvailable) {
    Serial.print("Mag x:");
    Serial.print(mx);
    Serial.print(" y:");
    Serial.print(my);
    Serial.print(" z:");
    Serial.print(mz);
    Serial.print(" | Gyro x:");
    Serial.print(gx);
    Serial.print(" y:");
    Serial.print(gy);
    Serial.print(" z:");
    Serial.print(gz);
    Serial.print(" | Acc x:");
    Serial.print(ax);
    Serial.print(" y:");
    Serial.print(ay);
    Serial.print(" z:");
    Serial.println(az);
  }

  // Add a delay to limit data rate
  delay(100);
}
